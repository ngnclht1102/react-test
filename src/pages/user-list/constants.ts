export const BASE_URL = 'https://jsonplaceholder.typicode.com';
export const GOOGLE_MAP_URL = 'https://maps.google.com/?q=';
