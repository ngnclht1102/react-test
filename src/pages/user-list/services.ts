import { BASE_URL } from './constants';
import { TUser } from './type';

export const Services = {
  getUsers: (): Promise<Array<TUser>> => {
    return new Promise((resolve, reject) => {
      fetch(`${BASE_URL}/users`)
        .then((response) => response.json())
        .then((json) => resolve(json))
        .catch((error) => reject(error));
    });
  },
};
