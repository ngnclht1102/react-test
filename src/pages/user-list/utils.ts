import { useCallback, useEffect, useState } from 'react';
import { GOOGLE_MAP_URL } from './constants';

export function useQuery<T>(promise: () => Promise<Array<T>>) {
  const [isIdle, setIsIdle] = useState<boolean>(true);
  const [isWorking, setIsWorking] = useState<boolean>(false);
  const [isError, setIsError] = useState<boolean>(false);
  const [errorMsg, setErrorMsg] = useState<string | undefined>();
  const [data, setData] = useState<Array<T>>([]);
  const [needRefreshAt, markAsNeedRefreshAt] = useState<number>();

  useEffect(() => {
    (async () => {
      setIsWorking(true);
      setIsError(false);
      setErrorMsg(undefined);
      try {
        const res = await promise();
        setData(res);
      } catch (error) {
        setIsError(true);
        setErrorMsg(error.message);
        setData([]);
      }
      setIsIdle(false);
      setIsWorking(false);
    })();
  }, [needRefreshAt, promise]);

  const refresh = useCallback(() => {
    setIsIdle(true);
    setData([]);
    markAsNeedRefreshAt(new Date().getMilliseconds());
  }, []);

  return {
    isIdle,
    isWorking,
    isError,
    errorMsg,
    data,
    refresh,
    needRefreshAt,
  };
}

export function getGoogleMapUrlByLatLng(lat: string, lng: string) {
  return `${GOOGLE_MAP_URL}${lat},${lng}`;
}
