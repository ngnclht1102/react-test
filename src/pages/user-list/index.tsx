import React from 'react';
import { useQuery } from '@pages/user-list/utils';
import { UserItem } from './components/user-item';
import { Services } from './services';
import { TUser } from './type';

export default (): JSX.Element => {
  const query = useQuery(Services.getUsers);
  if (query.isWorking || query.isIdle) {
    return <div id="loading">loading</div>;
  }
  if (query.isError) {
    return (
      <div id="error">
        <h1>Error while loading</h1>
        <button
          id="refresh-button"
          className="text-xl font-bold outline outline-offset-2 outline-1"
          onClick={query.refresh}
        >
          Refresh
        </button>
      </div>
    );
  }
  if (!query.data?.length && !query.isWorking && !query.isIdle) {
    return <div>Empty, there are no users.</div>;
  }

  return (
    <div className="container mx-auto w-4/5 my-10" id="container">
      <h1 className="text-xl font-bold">The list of users:</h1>
      <button
        id="refresh-button"
        className="text-xl font-bold outline outline-offset-2 outline-1"
        onClick={query.refresh}
      >
        Refresh
      </button>
      {query.data?.map((user: TUser) => {
        return <UserItem key={user.id} user={user} />;
      })}
    </div>
  );
};
