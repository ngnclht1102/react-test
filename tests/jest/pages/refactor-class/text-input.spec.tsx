import '@testing-library/jest-dom';
import { TUser } from '@pages/user-list/type';
import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import TextInput from '../../../../src/pages/refactor-class/components/input/index';

const formItem = {
  name: 'email',
  value: 'an existed value',
  placeholder: 'enter your email',
  onChange: (value: string) => {},
};
describe('Text input component', () => {
  // test 1: snapshot
  test('1. Should render component properly', () => {
    const componentRenderer = renderer.create(<TextInput {...formItem} type="text" />);
    const tree = componentRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // test 2
  test('2. Should render props correctly', () => {
    render(<TextInput {...formItem} type="text" />);
    expect(screen.getByTestId('wrapper')).toBeVisible();
    expect(screen.getByTestId('input')).toHaveAttribute('name', formItem.name);
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'text');
    expect(screen.getByTestId('input')).toHaveAttribute('placeholder', formItem.placeholder);
    expect(screen.getByTestId('input')).toHaveAttribute('value', formItem.value);
  });
});
