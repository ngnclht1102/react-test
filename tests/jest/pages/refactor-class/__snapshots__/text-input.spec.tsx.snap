// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text input component 1. Should render component properly 1`] = `
<div
  data-cy="textinput_email"
  data-testid="wrapper"
>
  <input
    data-testid="input"
    name="email"
    onChange={[Function]}
    placeholder="enter your email"
    type="text"
    value="an existed value"
  />
</div>
`;
